go/vt/vtgate/evalengine/fn_string.go

```diff
@@ -932,6 +937,105 @@ func (call builtinTrim) compile(c *compiler) (ctype, error) {
 	return ctype{Type: sqltypes.VarChar, Flag: flagNullable, Col: col}, nil
 }
 
+func (call *builtinSubstring) eval(env *ExpressionEnv) (eval, error) {
+	str, err := call.Arguments[0].eval(env)
+	if err != nil || str == nil {
+		return nil, err
+	}
+
+	text, ok := str.(*evalBytes)
+	tt := str.SQLType()
+	if !ok {
+		text, err = evalToVarchar(str, call.collate, true)
+		if err != nil {
+			return nil, err
+		}
+		tt = sqltypes.VarChar
+	}
+
+	p, err := call.Arguments[1].eval(env)
+	if err != nil || p == nil {
+		return nil, err
+	}
+
+	var l eval
+	if len(call.Arguments) > 2 {
+		l, err = call.Arguments[2].eval(env)
+		if err != nil || l == nil {
+			return nil, err
+		}
+	}
+
+	pos := evalToInt64(p).i
+	if pos == 0 {
+		return newEvalRaw(tt, nil, text.col), nil
+	}
+	cs := colldata.Lookup(text.col.Collation).Charset()
+	end := int64(charset.Length(cs, text.bytes))
+
+	if pos < 0 {
+		pos += end + 1
+	}
+	if pos < 1 || pos > end {
+		return newEvalRaw(tt, nil, text.col), nil
+	}
+
+	if len(call.Arguments) > 2 {
+		ll := evalToInt64(l).i
+		if ll < 1 {
+			return newEvalRaw(tt, nil, text.col), nil
+		}
+		if ll > end-pos+1 {
+			ll = end - pos + 1
+		}
+		end = pos + ll - 1
+	}
+	res := charset.Slice(cs, text.bytes, int(pos-1), int(end))
+	return newEvalRaw(tt, res, text.col), nil
+}
+
+func (call *builtinSubstring) compile(c *compiler) (ctype, error) {
+	str, err := call.Arguments[0].compile(c)
+	if err != nil {
+		return ctype{}, err
+	}
+
+	p, err := call.Arguments[1].compile(c)
+	if err != nil {
+		return ctype{}, err
+	}
+
+	tt := str.Type
+	skip1 := c.compileNullCheck2(str, p)
```

Comment:
Can move this closer to where it's actually used.

Reply:
No, that's actually not possible. It needs to emit the nil check in the generated code here, if that is done later it's wrong since we'd run other code before the nil check. This needs to happen before we run any conversions.